src/parser.ts

```diff
@@ -1870,6 +1871,44 @@ export class Parser extends DiagnosticEmitter {
     // implemented methods are virtual
     if (isInterface) flags |= CommonFlags.VIRTUAL;
 
+    var declareStart = 0;
+    var declareEnd = 0;
+    var contextIsAmbient = parent.is(CommonFlags.AMBIENT);
+    if (tn.peek() == Token.DECLARE) {
+      let state = tn.mark();
+      tn.next();
+      if (tn.peek() != Token.COLON) { // modifier
+        tn.discard(state);
+        if (isInterface) {
+          this.error(
+            DiagnosticCode._0_modifier_cannot_be_used_here,
+            tn.range(), "declare"
+          );
+        } else {
+          if (contextIsAmbient) {
+            this.error(
+              DiagnosticCode.A_declare_modifier_cannot_be_used_in_an_already_ambient_context,
+              tn.range()
+            ); // recoverable
+          } else {
+            this.error(
+              DiagnosticCode.Not_implemented_0,
+              tn.range(), "Ambient fields"
```

Comment:
Appears to also catch methods and properties, so perhaps:
```suggestion
              tn.range(), "Ambient class members"
```

Reply:
Just saw that there is another error below handling the case of non-fields, but seems one then gets both diagnosticss?

Reply:
Non-fields cannot be ambient, so ambient "class members" doesn't make sense. The logic is that one diagnostic will tell the user that ambient fields aren't supported, and one will tell the user that declaring non-fields is illegal :)

Reply:
Mostly thinking in terms of having a `declare someMethod();` that seems to produce both errors, with the first one being "Not implemented: Ambient fields" which seems odd. I would be fine with dropping the "Not implemented" one and just generally saying "'declare' modifier cannot be used here". Then the one below can be removed. Or am I missing something?

Reply:
What you're saying makes sense, but it is ultimately subjective. I'm relying on the fact that the user will see the other error and find that what they are doing is forbidden. "Ambient class members" just doesn't make sense to me as a concept because non-fields can't be ambient.

Reply:
The ideal solution would be of course to have exactly one diagnostic. For non-fields:

> 'declare' modifier cannot appear on class elements of this kind.(1031)

For fields:

> Not implemented: Ambient fields.

But two in the former case (the first referring to fields even though not declarring a field) seems odd :)